text/0000-named-arguments.md

```diff
@@ -0,0 +1,623 @@
+fn foo(.a _: i32) {}
+```
+
+Argument names can be added to positional arguments _from right to left_, without breaking backwards compatibility:
+
+```rust
+fn foo(a: i32, b: i32) {}
+fn foo(a: i32, .b: i32) {}
+fn foo(.a: i32, .b: i32) {}
+```
+
+Argument names can also be added to trait functions backwards compatibly:
+
+```rust
+trait Foo {
+    fn foo(a: i32, b: i32);
+}
+// -->
+trait Foo {
+    fn foo(.a: i32, .b: i32);
+}
+```
+
+## ABI compatibility
+
+Named arguments don't affect the ABI or code generation in any way. For example, you can use named arguments in `extern "C"` functions.
+
+## Interactions with the type system
+[interactions-with-the-type-system]: #interactions-with-the-type-system
+
+Named arguments don't interact with the type system, since named arguments are _not_ part of the function's type. For example, the following is invalid:
+
+```rust
+fn foo(.a: i32, .b: i32) {}
+
+let f: fn(.a: i32, .b: i32) = foo; // forbidden!
+let f: fn(i32, i32)         = foo; // correct
+
+fn higher_order(f: impl Fn(.a: i32, .b: i32)) {} // forbidden!
+fn higher_order(f: impl Fn(i32, i32)) {}         // correct
+```
+
+Also, named arguments are _not_ involved in type inference.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Minor: Verbosity
+
+In function calls, the syntax is slightly more verbose than in other languages such as Python or Swift:
+
+```rust
+// Swift
+greet(person: "Tim", alreadyGreeted: true)
+// Rust
+greet(.person = "Tim", .already_greeted = true)
+```
+
+## Minor: Named arguments aren't mandatory
+
+Because named arguments aren't mandatory in function calls, this might encourage bad habits where people don't specify named arguments out of laziness. This is especially bad when the function names are short and unspecific, as in the example above.
+
+However, this can also be seen as a good thing: If named arguments were mandatory, API authors might be hesitant to add named arguments, since they make function calls more verbose. However, if you can choose at every function call if you want to specify argument names, there are no downsides for an API author in making function arguments named.
+
+The main reason why named arguments aren't mandatory is that existing APIs can be converted to use named arguments backwards compatibly. This would not be possible if named arguments were mandatory.
+
+### Remedy: Add a clippy lint
+[remedy-add-a-clippy-lint]: #remedy-add-a-clippy-lint
+
+Laziness to specify named arguments can be remedied by adding a clippy lint that warns when a named argument is omitted.
+
+However, it should be allowed to omit the argument name, when it matches the variable, field or call expression that is passed as the argument, for example:
+
+```rust
+fn foo(.arg1: i32, .arg2: i32, .arg3: i32) {}
+
+let arg1 = 42;
+let s = Struct { arg2: 42 };
+fn arg3() -> i32 { 42 }
+
+foo(.arg1 = arg1, .arg2 = s.arg2, .arg3 = arg3()) // no warning
+foo(arg1, s.arg2, arg3())                         // no warning
+```
```

Comment:
This should belong to the function section. I am not sure, if it is very smart to add work to the compiler to check this.

Reply:
It's a lint, so it doesn't add work for the compiler, only to `cargo clippy`.